### DISCUSSION TITLE
How do I setup a remote worker?

### DISCUSSION THREAD

**Discussion by VBoB13** [apache/airflow#24275] [Q&A] [answered]
### Configuration
I have installed Airflow 2.3.1 on a virtual machine within our local network and I have installed the `apache-airflow[celery]` version.
I have run these commands to bring up the main application on the virtual machine:

1. `nohup airflow webserver $* >> ./logs/webserver.logs &`
2. `nohup airflow celery worker $* >> ./logs/worker.logs &`
3. `nohup airflow scheduler >> ./logs/scheduler.logs &`

And then I simply launch another worker on my local machine with:

- `nohup airflow celery worker --without-gossip $* >> ./logs/worker.logs &`

While it seems that I can execute the local machine's DAGs by invoking them through `airflow dags test [DAG ID] [BACKFILL DATE]`, I can't seem to be able to see these DAGs on the virtual machine's webserver. I am only able to see the DAGs that were put in the virtual machines designated DAG folder.

More details about my environment and log messages can be found [here (StackOverflow)](https://stackoverflow.com/questions/72525430/setup-airflow-with-remote-celery-worker).

Thanks in advance!

**Comment by potiuk** [comment, marked as answer]
You need to make your DAGs aare available in the same folder locally for all machines that need them (workers and scheduler) (i.e. your workers should see DAGs in the same folder as your scheduler). This can be achieved by a network filesystem or git-sync or similar.

**Reply by WraithCadmus** [reply]
Yes, don't forget to install any libs your DAGs need on the workers too. We put our DAGs in a git repo and pull them down, by the time the Scheduler refreshes it's all installed and ready to go.

**Reply by VBoB13** [reply]
Thank you so much! 🙇🏻‍♂️ This clarified some things I had suspected for some time 👍🏻

**Reply by potiuk** [reply]
ANd just to clarify - they do no have to be in "exactly same path" - that was a limitation in some of the previous versions but I think in recent versions of Airlfow it has been fixed - they have to be present in "exactly same RELATIVE path" to `AIRFLOW__CORE__DAGS_FOLDER` that can be different in different components.

**Reply by potiuk** [reply]
Yeah. The problem is that it's very difficult to "serialize" python code with ALL dependencies (utils, librariest etc.) and effectively you'd have to basically sync the whole folder between components somehoe if you would like to make it work in scale. Having shared volume is just using other tools that do the job better than airlfow would do internally (and in the way that is best for your deployment).

**Reply by VBoB13** [reply]
Does the same thing go for log files?
Because `airflow` does not seem to be able to retrieve/read the remote logs.
I have setup a shared filesystem where the DAGs now live, but I can't put my log files in there because of user rights (running `nohup airflow celery worker -l /share/logs/worker.logs --without-gossip $* >> /share/logs/worker.logs &` immediately spits out this error `-bash: /share/logs/worker.logs: Permission denied`).
I cannot seem to change the permissions and/or ownership of the mounted `/share` folder, thus I am looking for some workaround 😮‍💨

**Reply by WraithCadmus** [reply]
Logs are (at least in my pip3 install with most default settings) stored locally on the workers, and the frontend pulls them over 8793/tcp for display.

**Reply by potiuk** [reply]
Yep. @WraithCadmus - this is the way it shoudl work . I recommend to read docs about it https://airflow.apache.org/docs/apache-airflow/stable/logging-monitoring/logging-architecture.html

**Reply by VBoB13** [reply]
But when I try to launch a DAG that triggers my remote worker, the log link in the web UI shows this:
```
*** Log file does not exist: /home/lowi/airflow/logs/dag_id=lowi17/run_id=manual__2022-06-10T01:36:18.827448+00:00/task_id=動腦雜誌/attempt=1.log
*** Fetching from: http://:8793/log/dag_id=lowi17/run_id=manual__2022-06-10T01:36:18.827448+00:00/task_id=動腦雜誌/attempt=1.log
*** Failed to fetch log file from worker. Request URL is missing an 'http://' or 'https://' protocol.
```

What am I missing?

**UPDATE**

I just changed my `airflow.cfg` on my remote worker back to `base_log_folder=/home/{user}/airflow/logs` (same as main node) and the message that pops up changed to this:
```
*** Log file does not exist: /home/lowi/airflow/logs/dag_id=lowi17/run_id=manual__2022-06-10T02:07:25.573427+00:00/task_id=動腦雜誌/attempt=1.log
*** Fetching from: http://vm2:8793/log/dag_id=lowi17/run_id=manual__2022-06-10T02:07:25.573427+00:00/task_id=動腦雜誌/attempt=1.log
*** Failed to fetch log file from worker. [Errno -3] Temporary failure in name resolution
```

~~Is it maybe because of the Chinese names? Should it matter?~~
Since my other scripts on the main node works without the `Temporary failure in name resolution` error, it can't be...

But then I really don't know what's going on either 😵‍💫

**Reply by potiuk** [reply]
You need to make sure the https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#hostname-callable parameter point to a method that when called will return a host name that webserver shoudl be able to use when requesting logs. 

But what it is - it very much depend on the deployment you have

**Reply by potiuk** [reply]
For example you got 'vm2' as your worker name. an this is the hostname that is stored for the task. And this hostname is used by the webserver to reach out to the worker and connecti on 8793 port. Apparently - for a reason that only you can debug when webserver reached to `vm2` they got "tempporary failure in name resolution".  WHY ? HOW? NO IDEA. 

I just explained you step by step what the log file already told you. But you are the only person that can do something with that information.

**Reply by potiuk** [reply]
And in the first case you "hostname-callable" returned empty string. Why? How? No idea. Only you can tell.

And likely this is the right haiku for you at this stage:

![image](https://user-images.githubusercontent.com/595491/173117502-d9931642-450c-4cf0-9bd5-5f756ab01a37.png)

**Reply by VBoB13** [reply]
@potiuk "It's not DNS" -> ~~"There's no way it's DNS"~~ -> "It was DNS" Was the pathway my mind went :P
Jokes aside - I simply solved it by adding my own package + function that returns the local worker's IP. Now the logs all work as intended :)

Thank you very much for all the help! 🙇🏻‍♂️

**Reply by hieptran1812** [reply]
@VBoB13 How is your own package + function? I am having the same issues. My remote worker can not find the dags in main server.
